### ISSUE TITLE
cachedb_redis cachedb_raw_query

### ISSUE THREAD

**Issue by haeferer** [OpenSIPS/opensips#1250]
I've tried to store values containing  spaces, but this doesnt work. 
Is there a way to store values (and maybe keys) containing spaces?

Thx for your help and nice work 

Andreas

It'seems the code only works with space delemited parameters, so the query is not submitted "RAW", cause the attributes are splittet and then the query is reconstructed (and send to REDIS)

**Comment by bogdan-iancu** [comment]
@haeferer , please provide the exact script snippet that reflects the issue you are reporting.

**Comment by haeferer** [comment]
The Problem occurs if you send a RAW query like 

```
cache_raw_query("redis:cache","set callid:$avp(callid) $json(resultlog) EX 36000");
```

to set a key including EX Operations. The value needs to be without any space, cause
the libary does not send RAW, it identifies the value (as not containing spaces)

so the sample only works if the JSON does not have spaces or is encoded (our current solution) to base64
```
cache_raw_query("redis:cache","set callid:$avp(callid) $(json(resultlog){s.b64encode}) EX 36000");
```
(not good, but works)

**Comment by bogdan-iancu** [comment]
@haeferer , as OpenSIPS is dealing with this like a raw query, it has not idea what has to be skipped or not. So it is your duty, from script level, to properly format (as escape) the query you are pushing. Have you tried to escape the value with one of the escaping transformations - see http://www.opensips.org/Documentation/Script-Tran-2-3#toc9 ?

**Comment by haeferer** [comment]
Ok, i will check them. 

Thx

**Comment by haeferer** [comment]
I've tried the escape function to store JSON in Redis,

```
$json(my_server) := "{}";
$json(my_server/started) = $Tf;
cache_raw_query("redis:cache","set system:startup \"$(json(my_server){s.escape.common})\"");
```
the constructed redis-query (the string for the second parameter) is correct

```
set system:startup "{ \"started\": 1517565136 }"
```

The QueryResult is still not correct

**Comment by liviuchircu** [comment]
@haeferer this actually seems to be a limitation of the hiredis C client project.  We do nothing else but to pass the input string to the `redisvCommand()` API function:

```
Jul 10 16:35:38 [31878] INFO:cachedb_redis:redis_raw_query_send: shipping command to Redis: 'set system:startup "{ \"started\": 1562765738 }"'
```

... but the server immediately rejects it with:

```
Jul 10 16:35:38 [31878] ERROR:cachedb_redis:redis_raw_query_send: Redis operation failure - ERR syntax error
```

Surprisingly enough, the same syntax works if thrown via the `redis-cli`:

```
10.0.0.10:6379> set system:startup "{ \"started\": 1562765738 }"
OK
10.0.0.10:6379> GET system:startup
"{ \"started\": 1562765738 }"
```

So it seems you have to encode your JSONs into some contiguous strings before being able to store them, possibly using base64:

```
cache_raw_query("redis", "set system:startup $(json(my_server){s.b64encode})");
```

**Comment by liviuchircu** [comment]
A bit of a late update here: this "spaces" issue is now fixed starting with the `3.1.9` release, more details [here](https://github.com/OpenSIPS/opensips/issues/2036#issuecomment-1104174662).
